sycl/include/CL/sycl/detail/spec_const_integration.hpp

```diff
@@ -0,0 +1,31 @@
+//==---------------------- spec_const_integration.hpp ----------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+// This header file must not be included to any DPC++ headers.
+// This header file should only be included to integration footer.
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+
+#if __cplusplus >= 201703L
+// Translates SYCL 2020 specialization constant type to its name.
+template <auto &SpecName> const char *get_spec_constant_symbolic_ID() {
```

Comment:
Probably we can solve UB problem by simple adding a wrapper function, please see:
https://godbolt.org/z/5PhPKv5sx

@AlexeySachkov  @erichkeane WDYT?

Reply:
Yes, I believe that is the plan we had come up with in the meeting where we planned this feature, so I'm a touch surprised it didn't make it this far.

We need to make sure we don't instantiate get_spec-constant_symbolic_ID until the end (after the footer has made all its specializations).

In your example, you could just as easily skip the wrapper and delay the definition of `set_spec_constant` if you'd wanted, but I presume there is a good reason not to do that.

Reply:
> In your example, you could just as easily skip the wrapper and delay the definition of set_spec_constant if you'd wanted, but I presume there is a good reason not to do that.

I guess the main reason is code fragmentation, i.e. the class is defined in different places and those are not just `.hpp` + `.cpp` file.

The wrapper approach looks good to me

Reply:
@romanovvlad, @dm-vodopyanov, I've added this bit in b69f889e7241d09db8909931c2277536f97ff019, could you please take a look?